### DISCUSSION TITLE
Why can't `str | None | None` be assigned to `TypeVar(..., str, int, str | None, int | None)`

### DISCUSSION THREAD

**Discussion by adriangb** [microsoft/pyright#4735] [Q&A] [answered]
Here's the code in question:

```python
from dataclasses import dataclass
from typing import Any, Callable, Generic, TypeVar, overload

NewOutputT = TypeVar("NewOutputT", str, int, str | None, int | None)
InputT = TypeVar("InputT", str, int, str | None, int | None)
OutputT = TypeVar("OutputT", str, int, str | None, int | None)


class FieldType(Generic[InputT]):
    pass

@dataclass
class Mapper(Generic[InputT, OutputT]):
    chain: list[Callable[[Any], Any]] = []

    def map(
        self: "Mapper[InputT, OutputT]", f: Callable[[OutputT], NewOutputT]
    ) -> "Mapper[InputT, NewOutputT]":
        return Mapper([*self.chain, f])

    @overload
    def map_skipping_nulls(
        self: "Mapper[InputT, str]",
        f: Callable[[str], NewOutputT],
    ) -> "Mapper[InputT, NewOutputT]":
        ...

    @overload
    def map_skipping_nulls(
        self: "Mapper[InputT, int]",
        f: Callable[[int], NewOutputT],
    ) -> "Mapper[InputT, NewOutputT]":
        ...

    @overload
    def map_skipping_nulls(
        self: "Mapper[InputT, str | None]",
        f: Callable[[str], NewOutputT],
    ) -> "Mapper[InputT, NewOutputT | None]":
        ...

    @overload
    def map_skipping_nulls(
        self: "Mapper[InputT, int | None]",
        f: Callable[[int], NewOutputT],
    ) -> "Mapper[InputT, NewOutputT | None]":
        ...

    def map_skipping_nulls(
        self: "Mapper[Any, Any]", f: Callable[[Any], Any]
    ) -> "Mapper[Any, Any]":
        def wrapper(inp: Any) -> Any:
            if inp is None:
                return None
            return inp

        return Mapper([*self.chain, wrapper])

    def build(self: "Mapper[InputT, OutputT]") -> Callable[[InputT], OutputT]:
        def run(inp: InputT) -> OutputT:
            tf: Any = inp
            for f in self.chain:
                tf = f(tf)
            return tf

        return run

    @staticmethod
    def create(__inp: FieldType[NewOutputT]) -> "Mapper[NewOutputT, NewOutputT]":
        return Mapper()


# usage
def zero_to_null(v: int) -> int | None:
    return None if v == 0 else v


def divide_2_by(v: int) -> int:
    return 2 // v


def null_to_zero(v: int | None) -> int:
    return 0 if v is None else v


NullableIntField = FieldType[int | None]()

m = (
    Mapper.create(NullableIntField)
    .map_skipping_nulls(zero_to_null)
    .map_skipping_nulls(divide_2_by)
    .map(null_to_zero)
    .build()
)

v1: int = m(1)
v2: int = m(None)

```

The errors happen on line 39 and 46:

```
(variable) NewOutputT: Type[NewOutputT@map_skipping_nulls]
Type "NewOutputT@map_skipping_nulls | None" cannot be assigned to type variable "OutputT@Mapper"
  Type "NewOutputT@map_skipping_nulls | None" is incompatible with constrained type variable "OutputT@Mapper"
```

I'm not sure what the issue is: in my mind `NewOutputT` must be one of `{int, str, int | None, str | None}` and I am doing a union with `None` which I'd think results in `{int | None, str | None, int | None | None, str | None | None}` which then simplifies to `{int | None, str | None}` which should be assignable to `OutputT` since `OutputT` can be both `str | None` and `int | None`.

For what it's worth MyPy complains about the same thing:

```
demo.py:22: error: Overloaded function signatures 1 and 3 overlap with incompatible return types  [misc]
demo.py:29: error: Overloaded function signatures 2 and 4 overlap with incompatible return types  [misc]
demo.py:39: error: Invalid type argument value for "Mapper"  [type-var]
demo.py:46: error: Invalid type argument value for "Mapper"  [type-var]
```

(mypy also complains about overlapping overloads with incompatible return types. which seems related)

Neither mypy nor pyright complain about the _usage_: that part has no errors and even types correctly along the way and for the results, which is a bit surprising considering the errors in the class definition.

I'm guessing I'm not understanding something here. Any help would be much appreciated.

**Comment by erictraut** [comment, marked as answer]
You're receiving an error because you're specifying a type argument of `NewOutputT | None` to a type parameter that is invariant and constrained to one of the following: `str`, `int`, `str | None`, `int | None`. A constraint solver is going to map the `NewOutputT` portion of the type argument to all of these constraints, then it's going to be left with a `None` that doesn't map to any of the constraint types.

You're really abusing the type system here. It's unlikely this approach will work. I recommend rethinking your solution. I'm not sure what you're trying to implement, but I encourage you to avoid using constrained TypeVars if possible, and don't ever use unions in a TypeVar constraint.

**Reply by adriangb** [reply]
You're right this is an abuse of the type system. I'm just trying to type an existing thing in an untyped codebase, I have some other ideas I'll explore but they'd be more invasive of the codebase itself. Thank you for the feedback.

I am curious if you surprised at all that the implementation section works (and everything there typechecks correctly as a far as I can tell) despite the errors in the class definition?

> A constraint solver is going to map the NewOutputT portion of the type argument to all of these constraints, then it's going to be left with a None that doesn't map to any of the constraint types.

I have nearly zero knowledge of type constraint solvers but for what it's worth my intuition would have been that the solver would merge that union into `NewOutputT`'s constraints one by one and then check if the result satisfied all of `OutputT`'s constraints. Interestingly this does seem to work in some other contexts:

```python
T = TypeVar("T", int, str, int | None, str | None)

def foo(_: T) -> T | None:
    ...

foo(foo(1))
foo(foo("abc"))


def bar(_: T) -> T | float:
    ...

bar(bar(1))  # correctly fails
```

Which I would think does the same thing (assigns `T | None` to `T`).

**Reply by hmc-cs-mdrissi** [reply]
I'd generally avoid constrained type variables and prefer bound type variables. [Constrained type variables](https://github.com/python/typing/discussions/1080) are very python unique and were designed for str/bytes with strange issue that subclasses of str/bytes should be avoided. My guess what you actually want is,

`T = TypeVar("T", bound=int | str | None)`

There exist places where constrained type variables are right, but except for str/bytes situation they are esoteric.

**Reply by adriangb** [reply]
Mehdi thank you for pointing that out. I got led astray because Pandas is involved in this and [their type variables use constraints](https://github.com/pandas-dev/pandas-stubs/blob/6829a993697f8c5d65b721ca1cec3980ac79e69c/pandas-stubs/_typing.pyi#L298-L316C2) so for whatever reason I thought I had to as well.

Using type variables with a union bound makes everything work:

<details>
<summary>Code</summary>
<pre>


```python
from dataclasses import dataclass
from typing import Any, Callable, Generic, TypeVar, Union, overload

NewOutputT = TypeVar("NewOutputT", bound=Union[str, int, str | None, int | None])
InputT = TypeVar("InputT", bound=Union[str, int, str | None, int | None])
OutputT = TypeVar("OutputT", bound=Union[str, int, str | None, int | None])


class FieldType(Generic[InputT]):
    pass

@dataclass
class Mapper(Generic[InputT, OutputT]):
    chain: list[Callable[[Any], Any]] = []

    def map(
        self: "Mapper[InputT, OutputT]", f: Callable[[OutputT], NewOutputT]
    ) -> "Mapper[InputT, NewOutputT]":
        return Mapper([*self.chain, f])

    @overload
    def map_skipping_nulls(
        self: "Mapper[InputT, str]",
        f: Callable[[str], NewOutputT],
    ) -> "Mapper[InputT, NewOutputT]":
        ...

    @overload
    def map_skipping_nulls(
        self: "Mapper[InputT, int]",
        f: Callable[[int], NewOutputT],
    ) -> "Mapper[InputT, NewOutputT]":
        ...

    @overload
    def map_skipping_nulls(
        self: "Mapper[InputT, str | None]",
        f: Callable[[str], NewOutputT],
    ) -> "Mapper[InputT, NewOutputT | None]":
        ...

    @overload
    def map_skipping_nulls(
        self: "Mapper[InputT, int | None]",
        f: Callable[[int], NewOutputT],
    ) -> "Mapper[InputT, NewOutputT | None]":
        ...

    def map_skipping_nulls(
        self: "Mapper[Any, Any]", f: Callable[[Any], Any]
    ) -> "Mapper[Any, Any]":
        def wrapper(inp: Any) -> Any:
            if inp is None:
                return None
            return inp

        return Mapper([*self.chain, wrapper])

    def build(self: "Mapper[InputT, OutputT]") -> Callable[[InputT], OutputT]:
        def run(inp: InputT) -> OutputT:
            tf: Any = inp
            for f in self.chain:
                tf = f(tf)
            return tf

        return run

    @staticmethod
    def create(__inp: FieldType[NewOutputT]) -> "Mapper[NewOutputT, NewOutputT]":
        return Mapper()


# usage
def zero_to_null(v: int) -> int | None:
    return None if v == 0 else v


def divide_2_by(v: int) -> int:
    return 2 // v


def null_to_zero(v: int | None) -> int:
    return 0 if v is None else v


NullableIntField = FieldType[int | None]()

m1 = (
    Mapper.create(NullableIntField)
    .map_skipping_nulls(zero_to_null)
    .map_skipping_nulls(divide_2_by)
    .map(null_to_zero)
    .build()
)

m2 = (
    Mapper.create(NullableIntField)
    .map_skipping_nulls(zero_to_null)
    .map_skipping_nulls(divide_2_by)
    .build()
)

v1: int = m1(1)
v2: int = m1(None)

reveal_type(m2(1))  # int | None
reveal_type(m2(None))  # int | None
```

</pre>
</details>
